.github/workflows/build.yml

```diff
@@ -0,0 +1,163 @@
+        run: >
+          ctest --test-dir ./build -C ${{ matrix.build_type }} --output-on-failure
+          --exclude-regex "quickmaterialtest|quicktexturetest|connectiontest-style-filter|bindinginspectortest|quickinspectortest|quickinspectortest2|modelinspectortest"
+        env:
+          QT_QPA_PLATFORM: offscreen
+          QT_QUICK_BACKEND: software
+
+        # Exclude
+        #1 - connectiontest-preload (Failed)
+        #2 - connectiontest-preload-filter (Failed)
+        #3 - connectiontest-style (Failed)
+        #4 - connectiontest-style-filter (Failed)
+        #5 - connectiontest-lldb (Failed)
+        #6 - connectiontest-lldb-filter (Failed)
+        #23 - probeabidetectortest (Failed)
+        #26 - launchertest (Failed)
+        #37 - quickinspectortest2 (Failed)
+      - name: Run tests Qt5 on macOS
+        if: ${{ matrix.build_type == 'Debug' && runner.os == 'macOS' && startsWith(matrix.qt.version, '5.') }}
+        run: >
+          ctest --test-dir ./build -C ${{ matrix.build_type }} --output-on-failure
+          --exclude-regex "connectiontest-*|probeabidetectortest|launchertest|quickinspectortest2"
+
+        # Exclude
+        #1 - connectiontest-preload (Failed)
+        #2 - connectiontest-preload-filter (Failed)
+        #3 - connectiontest-style (Failed)
+        #4 - connectiontest-style-filter (Failed)
+        #5 - connectiontest-lldb (Failed)
+        #6 - connectiontest-lldb-filter (Failed)
+        #28 - probeabidetectortest (Failed)
+        #31 - launchertest (Failed)
+        #32 - clientconnectiontest (Failed)
+        # modelinspectortest fails in Qt6/CI passes locally
+        # quickinspectortest2
+      - name: Run tests Qt6 on macOS
+        if: ${{ matrix.build_type == 'Debug' && runner.os == 'macOS' && startsWith(matrix.qt.version, '6.') }}
+        run: >
+          ctest --test-dir ./build -C ${{ matrix.build_type }} --output-on-failure
+          --exclude-regex "connectiontest-*|probeabidetectortest|launchertest|clientconnectiontest|modelinspectortest|quickinspectortest2"
+
+        # Exclude
+        # quicktexturetest
+        # bindinginspectortest
+      - name: Qt5 Run tests on Windows
+        if: ${{ matrix.build_type == 'Debug' && runner.os == 'Windows' && startsWith(matrix.qt.version, '5.') }}
+        run: >
+          ctest --test-dir ./build -C ${{ matrix.build_type }} --output-on-failure
+          --exclude-regex "quicktexturetest|bindinginspectortest"
+
+        # Exclude
+        # quicktexturetest
+        # bindinginspectortest
+        # modelinspectortest fails in Qt6/CI passes locally
+      - name: Qt6 Run tests on Windows
+        if: ${{ matrix.build_type == 'Debug' && runner.os == 'Windows' && startsWith(matrix.qt.version, '6.') }}
+        run: >
+          ctest --test-dir ./build -C ${{ matrix.build_type }} --output-on-failure
+          --exclude-regex "quicktexturetest|launchertest|bindinginspectortest|modelinspectortest"
+
```

Comment:
This will _work_, no question. But except for the exclusion list, the test instances are basically identical. (99%, anyway — just the Linux envvars.)

IMHO it's cleaner to just use one definition, and break out the selection of exclusion list into a previous step. I got it to work like this, which is a little clunky but pretty readable...

```yaml
jobs:
  build:
# [...]
    env:
      Windows_5: 'quicktexturetest|bindinginspectortest'
      Windows_6: 'quicktexturetest|launchertest|bindinginspectortest|modelinspectortest'
      macOS_5: 'connectiontest-*|probeabidetectortest|launchertest|quickinspectortest2'
      macOS_6: 'connectiontest-*|probeabidetectortest|launchertest|clientconnectiontest|modelinspectortest|quickinspectortest2' 
      Linux_5: 'quickmaterialtest|quicktexturetest|connectiontest-style-filter'
      Linux_6: 'quickmaterialtest|quicktexturetest|connectiontest-style-filter|bindinginspectortest|quickinspectortest|quickinspectortest2|modelinspectortest'
      QT_QPA_PLATFORM: ${{ contains('ubuntu', matrix.os) && 'offscreen' || '' }}
      QT_QUICK_BACKEND: ${{ contains('ubuntu', matrix.os) && 'software' || '' }}
    steps:
      - name: "Select exclusions"
        env:
          CTEST_EXCLUSIONS: >
            ${{ (runner.os == 'Windows' && matrix.qt.version == 5) && env.Windows_5 || '' }}
            ${{ (runner.os == 'Windows' && matrix.qt.version == 6) && env.Windows_6 || '' }}
            ${{ (runner.os == 'Linux' && matrix.qt.version == 5) && env.Linux_5 || '' }}
            ${{ (runner.os == 'Linux' && matrix.qt.version == 6) && env.Linux_6 || '' }}
            ${{ (runner.os == 'macOS' && matrix.qt.version == 5) && env.macOS_5 || '' }}
            ${{ (runner.os == 'macOS' && matrix.qt.version == 6) && env.macOS_6 || '' }}
        run: >
          ${{ runner.os == 'windows'
          && '"CTEST_EXCLUSIONS=$env:CTEST_EXCLUSIONS" >> $env:GITHUB_ENV'
          || 'echo "CTEST_EXCLUSIONS=$CTEST_EXCLUSIONS" >> "$GITHUB_ENV"' }}
```

It defines all of the right envvars for each system, so they can be dropped in later (Including `--exclude-regex "${{ env.CTEST_EXCLUSIONS }}"` for the CTest invocation.)

Though now I'm also seeing that I have to restrict the tests to _non_-client-only builds. A bunch of them complain if they're missing a probe for the platform.

Reply:
(The above also relies on a `matrix.qt` setup of the form I suggested earlier:)

```yaml
matrix:
  qt:
    - version: 5
      request: 5.15.*
    - version: 6
      request: 6.4.*
      modules: qtshadertools
```

Reply:
(If you don't want to pollute the process environment, the exclusions could also be baked directly into the selection step, e.g.:)

```yaml
# [...]
    env:
      QT_QPA_PLATFORM: ${{ contains('ubuntu', matrix.os) && 'offscreen' || '' }}
      QT_QUICK_BACKEND: ${{ contains('ubuntu', matrix.os) && 'software' || '' }}
    steps:
      - name: "Select exclusions"
        env:
          CTEST_EXCLUSIONS: >
            ${{ (runner.os == 'Windows' && matrix.qt.version == 5)
            && 'quicktexturetest|bindinginspectortest' || '' }}
            ${{ (runner.os == 'Windows' && matrix.qt.version == 6) 
            && 'quicktexturetest|launchertest|bindinginspectortest|modelinspectortest' || '' }}
            ${{ (runner.os == 'Linux' && matrix.qt.version == 5) 
            && 'quickmaterialtest|quicktexturetest|connectiontest-style-filter' || '' }}
            ${{ (runner.os == 'Linux' && matrix.qt.version == 6) 
            && 'quickmaterialtest|quicktexturetest|connectiontest-style-filter|bindinginspectortest|quickinspectortest|quickinspectortest2|modelinspectortest' || '' }}
            ${{ (runner.os == 'macOS' && matrix.qt.version == 5) 
            && 'connectiontest-*|probeabidetectortest|launchertest|quickinspectortest2' || '' }}
            ${{ (runner.os == 'macOS' && matrix.qt.version == 6) 
            && 'connectiontest-*|probeabidetectortest|launchertest|clientconnectiontest|modelinspectortest|quickinspectortest2' || '' }}
        run: >
          ${{ runner.os == 'windows'
          && '"CTEST_EXCLUSIONS=$env:CTEST_EXCLUSIONS" >> $env:GITHUB_ENV'
          || 'echo "CTEST_EXCLUSIONS=$CTEST_EXCLUSIONS" >> "$GITHUB_ENV"' }}
```

That'll have the same effect, without defining all the `Linux_5`, `Windows_6`, etc. envvars that'd otherwise persist through the entire job run. The comments from the individual steps in the current workflow could also, **I _believe_**, be included between each condition without affecting the final result.

Reply:
Sorry but IMO it's a lot easier to have multiple steps that are way easier to read.
This workaround GitHub actions lack of ternary options is not obvious so I prefer to avoid it.